cmd/titus-storage/ebs.go

```diff
@@ -0,0 +1,272 @@
+			break
+		}
+		ctr++
+	}
+	return deviceName + string(runes[ctr]), nil
+}
+
+func doesDriveExist(driveName string) bool {
+	if _, err := os.Stat(driveName); os.IsNotExist(err) {
+		for _, file := range listNvmeBlockDevices() {
+			if d, _ := ebsnvme.ScanDevice(file); d.Name == driveName {
+				return true
+			}
+		}
+		return false
+	}
+	return true
+}
+
+func waitForDeviceToDisappear(driveName string, maxAttempts int) bool {
+	var attempts int
+	for doesDriveExist(driveName) {
+		time.Sleep(2 * time.Second)
+		attempts++
+		if attempts >= maxAttempts {
+			log.Printf("Waited the max %d seconds for %s to go, but it is still here!", attempts*2, driveName)
+			return false
+		}
+	}
+	log.Printf("Waited for %s to disappear, and it is now gone after %d seconds", driveName, attempts*2)
+	return true
+}
+
+func doesDriveExistWithTimeout(driveName string, maxAttempts int) bool {
+	var attempts int
+	for !doesDriveExist(driveName) {
+		time.Sleep(2 * time.Second)
+		attempts++
+		if attempts >= maxAttempts {
+			return false
+		}
+	}
+	return true
+}
+
+// GetActualBlockDeviceName returns the actual name of the block device seen
+// within the instance, like /dev/nvme1
+// This is often different from the ec2 device name, which is usually
+// something like /dev/sdx
+func GetActualBlockDeviceName(name string) (string, error) {
+	for _, device := range listNvmeBlockDevices() {
+		if d, _ := ebsnvme.ScanDevice(device); d.Name == name {
+			return device, nil
+		}
+	}
+	if _, err := os.Stat(name); os.IsNotExist(err) {
+		return "", err
+	}
+	return name, nil
+}
+
+func listNvmeBlockDevices() (devices []string) {
+	re := regexp.MustCompile(`(^\/dev\/nvme[0-9]+n1$)`)
+	f, _ := filepath.Glob("/dev/nvme*")
+	for _, d := range f {
+		if re.Match([]byte(d)) {
+			devices = append(devices, d)
+		}
+	}
+	return
+}
+
+// getDeviceName scans the Nvme on disk and looks at the
+// serial numbers of the devices and returns the device name
+// (/dev/nmve1). Behaves like a hash get, returning key, ok.
+// returning "", false if unfound.
+func getDeviceName(ebsVolumeID string) (string, bool) {
+	deviceName := ""
+	for _, device := range listNvmeBlockDevices() {
+		if d, _ := ebsnvme.ScanDevice(device); d.VolumeID == ebsVolumeID {
+			deviceName = d.Name
+		}
+	}
+	return deviceName, deviceName != ""
+}
+
+func attachEBS(ebsVolumeID string, ec2Client *ec2.EC2, instanceID string) (string, error) {
+	deviceName, ok := getDeviceName(ebsVolumeID)
+	if ok {
+		log.Printf("EBS Volume %s already attached as %s", ebsVolumeID, deviceName)
+		return deviceName, nil
+	}
+	deviceName, err := randDriveNamePicker()
+	if err != nil {
+		log.Fatal("Couldn't find an unused drive name")
+	}
+	attachVolIn := &ec2.AttachVolumeInput{
+		Device:     &deviceName,
+		InstanceId: &instanceID,
+		VolumeId:   &ebsVolumeID,
+	}
+	volAttachments, err := ec2Client.AttachVolume(attachVolIn)
+	if err != nil {
+		log.Fatalf("Couldn't attach: %v", err)
```

Comment:
Maybe return error here, there could be cleanup required in the caller?

Reply:
No cleanup required.
Kinda the beauty of this design is that the `stop` function can clean up anything, if required.